File: packages/browser/src/crawlers/reddit.ts
import { z } from "zod";
import { USER_AGENT, type Result, ok, err, createLogger } from "@repo/shared";

const logger = createLogger("crawler:reddit");

interface DiscoveredProduct {
  name: string;
  url: string;
  description: string;
  source: "reddit";
}

const redditListingSchema = z.object({
  data: z.object({
    children: z.array(
      z.object({
        data: z.object({
          title: z.string(),
          selftext: z.string(),
          url: z.string(),
          is_self: z.boolean(),
          domain: z.string(),
        }),
      })
    ),
  }),
});

const SUBREDDITS = [
  "https://www.reddit.com/r/SaaS/new.json?limit=25",
  "https://www.reddit.com/r/startups/new.json?limit=25",
];

const SKIP_DOMAINS = new Set([
  "reddit.com",
  "i.redd.it",
  "v.redd.it",
  "imgur.com",
  "youtube.com",
  "youtu.be",
  "twitter.com",
  "x.com",
]);

function extractUrls(text: string): string[] {
  const urlRegex = /https?:\/\/[^\s)>\]]+/g;
  return (text.match(urlRegex) ?? []).filter((url) => {
    try {
      const host = new URL(url).hostname.replace("www.", "");
      return !SKIP_DOMAINS.has(host);
    } catch {
      return false;
    }
  });
}

async function fetchSubreddit(url: string): Promise<DiscoveredProduct[]> {
  const res = await fetch(url, {
    headers: { "User-Agent": USER_AGENT },
  });

  if (!res.ok) {
    logger.warn("Reddit fetch failed", { url, status: res.status });
    return [];
  }

  const json: unknown = await res.json();
  const parsed = redditListingSchema.safeParse(json);

  if (!parsed.success) {
    logger.warn("Invalid Reddit response", { url, error: parsed.error.message });
    return [];
  }

  const results: DiscoveredProduct[] = [];

  for (const post of parsed.data.data.children) {
    const { title, selftext, url: postUrl, is_self, domain } = post.data;

    // External link posts
    if (!is_self && !SKIP_DOMAINS.has(domain.replace("www.", ""))) {
      results.push({
        name: title.slice(0, 200),
        url: postUrl,
        description: selftext.slice(0, 500),
        source: "reddit",
      });
      continue;
    }

    // Self posts — extract URLs from body
    if (is_self && selftext) {
      const urls = extractUrls(selftext);
      if (urls.length > 0) {
        results.push({
          name: title.slice(0, 200),
          url: urls[0],
          description: selftext.slice(0, 500),
          source: "reddit",
        });
      }
    }
  }

  return results;
}

export async function discoverFromReddit(): Promise<Result<DiscoveredProduct[]>> {
  const results = await Promise.all(SUBREDDITS.map(fetchSubreddit));
  return ok(results.flat().slice(0, 20));
}

File: apps/go-api/internal/browser/crawlers/reddit.go
package crawlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
)

// skipDomains are domains we filter out — not SaaS product sites.
var skipDomains = map[string]struct{}{
	"reddit.com":    {},
	"imgur.com":     {},
	"youtube.com":   {},
	"youtu.be":      {},
	"twitter.com":   {},
	"x.com":         {},
	"github.com":    {},
	"gist.github.com": {},
	"i.redd.it":     {},
	"v.redd.it":     {},
}

var urlRegex = regexp.MustCompile(`https?://[^\s\)\]"'<>]+`)

type redditListing struct {
	Data struct {
		Children []struct {
			Data struct {
				Title    string `json:"title"`
				URL      string `json:"url"`
				Selftext string `json:"selftext"`
				Domain   string `json:"domain"`
			} `json:"data"`
		} `json:"children"`
	} `json:"data"`
}

// DiscoverFromReddit fetches recent posts from r/SaaS and r/startups using
// Reddit's public JSON API and extracts product URLs.
func DiscoverFromReddit() ([]DiscoveredProduct, error) {
	subreddits := []string{"SaaS", "startups"}
	var products []DiscoveredProduct

	client := &http.Client{Timeout: 15 * time.Second}

	for _, sub := range subreddits {
		url := fmt.Sprintf("https://www.reddit.com/r/%s/new.json?limit=50", sub)
		req, err := http.NewRequest(http.MethodGet, url, nil)
		if err != nil {
			return nil, fmt.Errorf("creating request for r/%s: %w", sub, err)
		}
		req.Header.Set("User-Agent", "AgenticReviewer/1.0")

		resp, err := client.Do(req)
		if err != nil {
			log.Warn("reddit fetch failed", "subreddit", sub, "error", err)
			continue
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			log.Warn("reddit returned non-200", "subreddit", sub, "status", resp.StatusCode)
			continue
		}

		var listing redditListing
		if err := json.Unmarshal(body, &listing); err != nil {
			log.Warn("reddit json parse failed", "subreddit", sub, "error", err)
			continue
		}

		for _, child := range listing.Data.Children {
			post := child.Data

			// Try the post URL first.
			if post.URL != "" && !isSkipDomain(post.Domain) && !strings.Contains(post.URL, "reddit.com") {
				products = append(products, DiscoveredProduct{
					Name:        truncate(post.Title, 100),
					URL:         post.URL,
					Description: truncate(post.Selftext, 300),
					Source:      "reddit",
				})
				continue
			}

			// Extract URLs from self posts.
			if post.Selftext != "" {
				urls := urlRegex.FindAllString(post.Selftext, -1)
				for _, u := range urls {
					if !isSkipURL(u) {
						products = append(products, DiscoveredProduct{
							Name:        truncate(post.Title, 100),
							URL:         u,
							Description: truncate(post.Selftext, 300),
							Source:      "reddit",
						})
						break // Take only the first valid URL per post.
					}
				}
			}
		}
	}

	return products, nil
}

func isSkipDomain(domain string) bool {
	domain = strings.TrimPrefix(domain, "www.")
	_, skip := skipDomains[domain]
	return skip
}

func isSkipURL(rawURL string) bool {
	for domain := range skipDomains {
		if strings.Contains(rawURL, domain) {
			return true
		}
	}
	return false
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen]
}
